docs/ldml/tr35-keyboards.md

```diff
@@ -308,6 +308,336 @@ For extensibility, the `<special>` element will be allowed at nearly every level
+        } else if(hasNFDBoundaryBefore(i) && !have_marker) {
+            segmentEnd = i; // end of segment
+            i++; // move past the NFD char
+        } else if(have_marker) {
+            i += /* marker */; // skip past marker
+        } else {
+            // non boundary, just move index forward
+            i++;
+        }
+
+        if (segmentStart != segmentEnd) {
+            const segment    = input.substring(segmentStart, segmentEnd);
+            const segmentNfd = nfdMarkersSegment(segment);
+            output.append(segmentNfd);
+            segmentStart = segmentEnd;
+        }
+    } while(segmentEnd != input.end);
+
+    return output;
+}
+
+/**
+ * @param input segment text
+ * @returns normalized output segment
+ */
+function nfdMarkersSegment(input /* string with markers */) {
+    const noMarkers    = removeMarkers(input);
+    const markerMap    = parseMarkers(input);
+    const noMarkersNfd = normalizeNFD(noMarkers);
+    if (markerMap.empty()) {
+        // no markers in this segment.
+        return noMarkersNfd;
+    } else if(noMarkersNfd == noMarkers) {
+        // no change in this segment
+        return input;
+    } else {
+        return addBackMarkers(noMarkersNfd, markerMap);
+    }
+}
+
+// represents a removed marker, to be re-added
+interface MarkerEntry {
+    glue: string;      // 'glue' character
+    marker: Marker;  // which marker
+    processed: boolean = false; // track whether this was processed or not
+    eot: boolean = false; // track whether end of text or not
+}
+
+/** parse the string for markers */
+function parseMarkers(s: string): MarkerEntry[] {
+    // Note: parseMarkers() and removeMarkers() (not shown) could be the same function.
+    // queue of markers found
+    let lastMarkers: Marker[] = [];
+    // return value
+    let outputMap: MarkerEntry[] = [];
+
+
+    // iterate string by codepoints
+    for (i = s.begin; i< s.end;) {
+        if (markerAt(i)) {
+            // found a marker - add it to the list
+            lastMarkers.add(markerAt(i));
+            i += /* marker length */;
+        } else {
+            // From 'i' find the glue char for all markers found so far
+            // (in order)
+            // use first char of decomposed string
+            const glue = normalizeNFD(i).codePointAt(0);
+            for (m of lastMarkers) {
+                outputMap.add({marker: m, glue: glue});
+            }
+            lastMarkers.clear();
+
+            i++; // move to next codepoint
+        }
+    }
+
+    // ran out of codepoints.
+    // any remaining markers are EOT
+    for (m of lastMarkers) {
+        outputMap.add({marker: m, eot: true});
+    } 
+}
+
+/**
+ * re-add markers to input. Mutates markerMap.
+ * @param input normalized string to re-add markers
+ * @param markerMap list of glued markers. will be mutated
+ */
+function addBackMarkers(input : string, map: MarkerEntry[]) : stringWithMarkers {
```

Comment:
When would you want to re-add markers?

Reply:
It's called by one of the other functions.